docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,251 @@
+|>Sort Tables                        |  9,1339|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Write Header of tablebuf           |  9,1430|  0,0001|    218|      0|      0|      0|    457|     46|
+|>Write Tables to tablebuf           |  9,1503|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Layout Metadata                    |  9,1588|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Write Metadata Header              |  9,1656|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Write Metadata Tables              |  9,1731|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Write Metadata Strings             |  9,1819|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Write Metadata User Strings        |  9,1897|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Write Blob Stream                  |  9,1976|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Fixup Metadata                     |  9,2033|  0,0002|    218|      0|      0|      0|    457|     46|
+|>Generated IL and metadata          |  9,2102|  0,0015|    218|      0|      0|      0|    457|     46|
+|>Layout image                       |  9,2195|  0,0025|    218|      0|      0|      0|    457|     46|
+|>Writing Image                      |  9,2272|  0,0006|    218|      0|      0|      0|    456|     46|
+|>Signing Image                      |  9,2338|  0,0000|    218|      0|      0|      0|    456|     46|
+// main6
+|Write .NET Binary                   |  9,2463|  0,1930|    218|      0|      0|      0|    464|     46|     <-- long
+--------------------------------------------------------------------------------------------------------
+```
+
+The phases taking the longest are marked in the right. Those are assembly import, type check, optimization and IL writing.
+
+## Implementation plan
+
+The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages where each stage brings gains in some scenarios.
+
+**Stage 1 - force-gen and compare typechecking graphs**
+
+We already create the TC graph in `main1` and we are able to dump it (in Mermaid) via the compiler flags.
+
+So we can do the following:
+- force-gen and save the graph
+- append all the extra compilation info (basically, the compilation arguments string)
+- before doing the retypecheck, detect if there is any change to above:
+  - if so, skip the retypecheck
+  - otherwise, apply some algorithm to get the graph diff. The result should be the list of files for retypecheck (just ignore it for now)
+
+This will allow to skip a huge part of the compilation in the scenarios like reopening Visual Studio for a project when nothing has changed.
```

Comment:
```suggestion
This will allow to skip a huge part of the compilation in the scenarios like reopening Visual Studio for a project when nothing has changed.


This step is largely technical without big observable benefits, but will require the necessary msbuild hooks to communicate the intermmediatefiles folder, add time-based and hash-based cache invalidation logic, and involve necessary tests which will include the `clean` and `rebuild` tests to make sure the cache be easily invalidated on demand.
```